Awards query

thomasplevy opened the pull request:
## Description

Implements the `LLMS_Awards_Query` class, a wrapper around `WP_Query` for querying information about awarded certificates and achievments.

This is a (mostly) complete WIP but I want some feedback before going much further.

Also, I thought you might want to see if this will work for what you're working on with syncing @eri-trabiccolo and see if we need to add or change anything for your needs / requirements there.

I've been struggling to figure out the best way to replace the existing student query methods for certs and achievements. This includes the best I've come up with which would be to *not touch* the existing default behavior and have an alternate method signature allowing an array of `LLMS_Awards_Query` arguments to be passed into it for the future (new) behavior.

This would allow developers to opt into the new behavior without us having to figure out how to push the existing behavior into our new (more powerful) query.

What do you think about this strategy @eri-trabiccolo ?

## How has this been tested?

+ New unit tests

## Screenshots <!-- if applicable -->

## Types of changes

+ New feature
+ Deprecation for existing student `get_certificate()` and later `get_achievements()` methods.

## Checklist:
- [ ] My code has been tested.
- [ ] My code passes all existing automated tests. <!-- Check code: `composer run-script tests-run`, Guidelines: https://github.com/gocodebox/lifterlms/blob/trunk/tests/README.md -->
- [ ] My code follows the LifterLMS Coding & Documentation Standards. <!-- Check code: `composer run-script check-cs-errors`, Guidelines: https://github.com/gocodebox/lifterlms/blob/trunk/docs/coding-standards.md and https://github.com/gocodebox/lifterlms/blob/trunk/docs/documentation-standards.md -->

thomasplevy committed:
update default images, per #1081

thomasplevy committed:
Refactor query abstract, add awards query class

thomasplevy committed:
Add awards query class

thomasplevy committed:
Try certificate deprecation

thomasplevy committed:
Fix headers

eri-trabiccolo commented on includes/class-llms-awards-query.php:
```diff
@@ -0,0 +1,526 @@
+	 *
+	 * @var null
+	 */
+	public $query = null;
```
I would make these `private` (for now, and `protected` if we think to extend it in the near future), and provide `getters`.

eri-trabiccolo commented on includes/class-llms-awards-query.php:
```diff
@@ -0,0 +1,526 @@
+<?php
+/**
+ * LLMS_Awards_Query class file
```
```suggestion
 * LLMS_Awards_Query class file.
```

eri-trabiccolo commented on includes/class-llms-awards-query.php:
```diff
@@ -0,0 +1,526 @@
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Query awarded achievements and engagements
```
```suggestion
 * Query awarded achievements and engagements.
```

eri-trabiccolo commented on includes/class-llms-awards-query.php:
```diff
@@ -0,0 +1,526 @@
+	public $query = null;
+
+	/**
+	 * Constructor
```
```suggestion
	 * Constructor.
```

eri-trabiccolo commented on includes/class-llms-awards-query.php:
```diff
@@ -0,0 +1,526 @@
+	 *     @type int       $per_page               Number of results to display per page. Use `-1` to show all possible results.
+	 *     @type boolean   $no_found_rows          Whether to skip counting the total rows found.
+	 * }
+	 */
```
```suggestion
	 * @return void
	 */
```

eri-trabiccolo commented on includes/class-llms-awards-query.php:
```diff
@@ -0,0 +1,526 @@
+	 * @since [version]
+	 *
+	 * @param string $type       Award type to query. Accepts: "achievements" or "certificates".
+	 * @param array  $query_args {
```
```suggestion
	 * @param array  $args {
```

eri-trabiccolo commented on includes/class-llms-awards-query.php:
```diff
@@ -0,0 +1,526 @@
+	 *     @type int|int[] $templates              Include awards created from the specified `llms_achievement` or `llms_certificate` template post(s) by WP_Post ID.
+	 *     @type int|int[] $templates__exclude     Exclude awards created from the specified `llms_achievement` or `llms_certificate` template  post(s) by WP_Post ID.
+	 *     @type boolean   $manual                 Include only awards created manually. If specified the `$related_posts`, `$related_posts__exclude`, `$engagements`, `$engagements__exclude`, `$templates`, and `$templates__exclude` arguments will be ignored.
+ 	 *     @type int       $page                   Results page number.
```
```suggestion
	 *     @type int       $page                   Results page number.
```

eri-trabiccolo commented on includes/class-llms-awards-query.php:
```diff
@@ -0,0 +1,526 @@
+		/**
+		 * Filters the allowed sort fields.
+		 *
+		 * They dynamic portion of this hook, `$this->type`, refers to the award type
```
```suggestion
		 * The dynamic portion of this hook, `$this->type`, refers to the award type
```

eri-trabiccolo commented on includes/class-llms-awards-query.php:
```diff
@@ -0,0 +1,526 @@
+			if ( ! in_array( $field, $this->get_allowed_sort_fields(), true ) ) {
+				unset( $sort[ $field ] );
+			}
+
```
```suggestion
```

eri-trabiccolo commented on includes/class-llms-awards-query.php:
```diff
@@ -0,0 +1,526 @@
+		 *
+		 * @since [version]
+		 *
+		 * @param object  $object     The retrieve object.
```
```suggestion
		 * @param object  $object     The retrieved object.
```

eri-trabiccolo commented on includes/class-llms-awards-query.php:
```diff
@@ -0,0 +1,526 @@
+	 * @return boolean
+	 */
+	public function has_results() {
+		return count( $this->query->posts ) >= 1;
```
```suggestion
		return count( $this->query->posts ) > 0;
```
golf :P

eri-trabiccolo commented on includes/class-llms-awards-query.php:
```diff
@@ -0,0 +1,526 @@
+					),
+				);
+			}
+
```
```suggestion
```

eri-trabiccolo commented on includes/models/model.llms.student.php:
```diff
@@ -288,19 +288,39 @@ public function get_enrollment_order( $product_id ) {
+	 * @param string       $order           Deprecated signature only: Ordering method for returned results (ASC or DESC).
+	 * @param string       $return          Deprecated signature only: Return type. Accepts "obj" for an array of objects from
+	 *                                      $wpdb->get_results and "certificates" for an array of LLMS_User_Certificate instances.
+	 * @return LLMS_Awards_Query|object[]|LLMS_User_Certificate[]
```
I'm not sure about returning the query to be honest, I think this method should still return the certificates, either the WP_Posts or the LLMS_User_Certificate objects.
Let's introduce `get_certificates_query()` if we want to return the query object.
What do you think?

thomasplevy replied:
Mmmm... I don't know... if we do that then we should probably have the default arguments to be `per_page=-1` which despite it being how this is working currently (before the change) doesn't seem like the best way to do this in general. For simplicities sake my plan was to change the places we use this (to the new behavior, and pass in `per_page=-1`) but we know that doing that is underperformant and that we should be paginating results to 10 - 20 per page anyway.

You know what I mean?

eri-trabiccolo replied:
Yeah but really that return hurts my brain :D

eri-trabiccolo commented on includes/class-llms-awards-query.php:
```diff
@@ -0,0 +1,526 @@
+			'page'                   => 1,
+			'per_page'               => 10,
+			'no_found_rows'          => false,
+		);
```
can we also have the `fields` arg?
https://developer.wordpress.org/reference/classes/wp_query/#return-fields-parameter
limited to `all` and `ids`.
Also, when this arg is set to `ids` the `get_results()` method below should always return a list of ids, no matter what you feed it with.
What do you think?
A typical case would be when I only want to count the number of awarded certificates, for performance reasons I'm not interested in anything else than the ids...

thomasplevy replied:
`fields` is now available via `LLMS_Abstract_Posts_Query`

Commit:
Update includes/class-llms-awards-query.php

Co-authored-by: Rocco Aliberti <raliberti84@gmail.com>

Commit:
Apply suggestions from code review

Co-authored-by: Rocco Aliberti <raliberti84@gmail.com>

thomasplevy committed:
:wq
erge branch 'engagement-refactoring' of github.com:gocodebox/lifterlms into awards-query

thomasplevy committed:
Merge branch 'awards-query' of github.com:gocodebox/lifterlms into awards-query

eri-trabiccolo commented on includes/class-llms-awards-query.php:
```diff
@@ -0,0 +1,525 @@
+	public function has_results() {
+		return count( $this->query->posts ) > 0;
+	}
+
```
Maybe we can add this useful method too:
https://github.com/gocodebox/lifterlms/blob/e9e27855088754b27b2fc2010fe69260327eb5a2/includes/abstracts/abstract.llms.database.query.php#L268

thomasplevy replied:
I started merging the query classes yesterday. I'm not finished yet

eri-trabiccolo replied:
I know, I'm giving you feedbacks while I use it <3

eri-trabiccolo commented on includes/class-llms-awards-query.php:
```diff
@@ -0,0 +1,525 @@
+	 *     @type int|int[] $templates__exclude     Exclude awards created from the specified `llms_achievement` or `llms_certificate` template  post(s) by WP_Post ID.
+	 *     @type boolean   $manual                 Include only awards created manually. If specified the `$related_posts`, `$related_posts__exclude`, `$engagements`, `$engagements__exclude`, `$templates`, and `$templates__exclude` arguments will be ignored.
+	 *     @type int       $page                   Results page number.
+	 *     @type int       $per_page               Number of results to display per page. Use `-1` to show all possible results.
```
The actual WP_Query arg is `posts_per_page`.

thomasplevy replied:
This isn't passed directly into WP_Query, but thank you.

eri-trabiccolo commented on includes/class-llms-awards-query.php:
```diff
@@ -0,0 +1,525 @@
+			'per_page'       => $args['per_page'],
+			'no_found_rows'  => $args['no_found_rows'],
+			'orderby'        => $args['sort'],
+		);
```
We need a `post_status` arg too:
https://developer.wordpress.org/reference/classes/wp_query/#status-parameters

Basically when not in admin the `post_status` defaults to only `publish` but we want to fetch also scheduled awarded certificates: we're not in admin when doing background processors.

thomasplevy replied:
Available via `LLMS_Abstract_Posts_Query`

eri-trabiccolo replied:
yea I saw, the `status` arg.

thomasplevy committed:
Add an escape and quote string utility function, per #1027

thomasplevy committed:
CS

thomasplevy committed:
Add a type option for award type

thomasplevy committed:
Add alternate signature

thomasplevy committed:
Fix deprecations

thomasplevy committed:
Stop accessing query properties directly

eri-trabiccolo commented on includes/abstracts/llms-abstract-posts-query.php:
```diff
@@ -0,0 +1,232 @@
+
+		$this->number_results = $this->wp_query->post_count;
+		$this->found_results  = $this->found_results();
+		$this->max_pages      = $this->wp_query->max_num_pages;
```
```suggestion
		$this->max_pages      = (int) $this->wp_query->max_num_pages;
```
Since `$this->wp_query->max_num_pages` seems to be a `float`, let's cast this (for future comparisons, e.g. in `is_last_page()` method).

thomasplevy commented:
@eri-trabiccolo this isn't completely ready but I think it's good to start looking at or working with. I'm working through updating all places in the codebase that deprecation warnings are being thrown as a result of these changes... Once again my refactoring has gotten out of control. I'm sorry.

eri-trabiccolo commented:
@thomasplevy I'm already working with this in this PR https://github.com/gocodebox/lifterlms/pull/1857
and works fine, for that purpose, except for this: https://github.com/gocodebox/lifterlms/pull/1859#discussion_r752104190

My PR is not rebased against this branch though, if you agree I'm going to do it right now.

thomasplevy commented:
@eri-trabiccolo Added your suggestion!

eri-trabiccolo commented on includes/abstracts/abstract.llms.database.query.php:
```diff
@@ -17,7 +17,7 @@
  * @since 3.34.0 Sanitizes sort parameters.
  */
-abstract class LLMS_Database_Query {
+abstract class LLMS_Database_Query extends LLMS_Abstract_Query {
```
file header changelog (meaning `@version [version]`)

eri-trabiccolo commented on includes/abstracts/abstract.llms.database.query.php:
```diff
@@ -27,158 +27,69 @@ abstract class LLMS_Database_Query {
+			return $this->default_arguments();
 		}
 
+		// Get the from the parent with the new replacement filter.
```
```suggestion
		// Get them from the parent with the new replacement filter.
```

eri-trabiccolo commented on includes/abstracts/abstract.llms.database.query.php:
```diff
@@ -27,158 +27,69 @@ abstract class LLMS_Database_Query {
 		 * @param array $args Array of default arguments to set up the query with.
 		 */
-		return apply_filters( 'llms_db_query_get_default_args', $args );
+		return apply_filters_deprecated( 'llms_db_query_get_default_args', array( $args ), '[version]', 'llms_{$this->id}_query_get_default_args' );
```
```suggestion
		return apply_filters_deprecated( 'llms_db_query_get_default_args', array( $args ), '[version]', "llms_{$this->id}_query_get_default_args" );
```
did you mean this, right?

thomasplevy replied:
I didn't, because I was thinking since it was a dynamic filter/string it should output as such but that doesn't make sense actually, since the dev using the filter would need to lookup the id instead of us just giving it to them.

eri-trabiccolo commented on includes/abstracts/abstract.llms.database.query.php:
```diff
@@ -235,220 +119,56 @@ protected function get_skip() {
 	 *
-	 * @since 3.8.0
-	 * @since 4.5.1 Drop use of `$this->get_filter('prepare_query')` in favor of `"llms_{$this->id}_query_prepare_query"`.
+	 * @see [Reference]
+	 * @link [URL]
 	 *
```
```suggestion
```

eri-trabiccolo commented on includes/abstracts/abstract.llms.database.query.php:
```diff
@@ -235,220 +119,56 @@ protected function get_skip() {
+	 * @link [URL]
 	 *
-	 * @return void
+	 * @return [type] [description]
```
```suggestion
	 * @return array An integer-keyed array of row objects.
```

eri-trabiccolo commented on includes/abstracts/abstract.llms.database.query.php:
```diff
@@ -235,220 +119,56 @@ protected function get_skip() {
-
-		$this->sanitize_sort();
+		global $wpdb;
+		return absint( $wpdb->get_var( 'SELECT FOUND_ROWS()' ) ); // db call ok; no-cache ok.
```
```suggestion
		return (int) $wpdb->get_var( 'SELECT FOUND_ROWS()' ); // db call ok; no-cache ok.
```
can't we just cast this to int? Are we worried about negative values?

thomasplevy replied:
We have different habits. I feel this *barely* matters but I'll do as you suggest because I don't feel strong enough about it one way or another.

eri-trabiccolo commented on includes/abstracts/abstract.llms.database.query.php:
```diff
@@ -556,4 +276,142 @@ protected function sql_orderby() {
+	 */
+	protected function prepare_query() {
+		if ( method_exists( $this, 'preprare_query' ) ) {
+			$class = get_called_class();
+			// _deprecated_function( "{$class}::preprare_query()", '[version]', "{$class}::prepare_query()" );
```
```suggestion
```
right?

thomasplevy replied:
I was arguing with myself about whether to update the whole codebase. I commented it out to test things without the warning throwing so I could see if it was working properly. I'm going to restore it later.

eri-trabiccolo commented on includes/abstracts/abstract.llms.database.query.php:
```diff
@@ -556,4 +276,142 @@ protected function sql_orderby() {
+			// _deprecated_function( "{$class}::preprare_query()", '[version]', "{$class}::prepare_query()" );
+			return $this->preprare_query();
+		}
+		// else {
+		// _doing_it_wrong(
+		// __METHOD__,
+		// sprintf( __( "Method '%s' not implemented. Must be overridden in subclass.", 'lifterlms' ), __METHOD__ )
+		// '[version]'
+		// );
+		// }
```
```suggestion
```

thomasplevy replied:
Same as above

eri-trabiccolo commented on includes/abstracts/llms-abstract-posts-query.php:
```diff
@@ -0,0 +1,232 @@
+<?php
+/**
+ * LLMS_Abstract_Posts_Query class file
```
```suggestion
 * LLMS_Abstract_Posts_Query class file.
```

eri-trabiccolo commented on includes/abstracts/llms-abstract-posts-query.php:
```diff
@@ -0,0 +1,232 @@
+			'per_page'   => 'posts_per_page',
+			'post_types' => 'post_type',
+			'search'     => 's',
+			'sort'       => 'orderby',
```
not sue, in the abstract `sort` is used bot for `order` and `orderby`, what happens here?

thomasplevy replied:
WP_Query `orderby` supports an array of key=>values where key is "orderby" and value is "order" -- so you can pass:

```php
'orderby' => array( 'date' => 'DESC', 'ID' => 'DESC' )
```

instead of:

```
'orderby' => 'date', 'id',
'order' => 'DESC'
```

This is buried in the parameter dosc: https://developer.wordpress.org/reference/classes/wp_query/#order-orderby-parameters

eri-trabiccolo replied:
My bad and ignorance! <3

eri-trabiccolo commented on includes/abstracts/llms-abstract-query.php:
```diff
@@ -0,0 +1,589 @@
+<?php
+/**
+ * LLMS_Abstract_Query class file
```
```suggestion
 * LLMS_Abstract_Query class file.
```

eri-trabiccolo commented on includes/abstracts/llms-abstract-query.php:
```diff
@@ -0,0 +1,589 @@
+	/**
+	 * Default arguments before merging with original.
+	 *
+	 * @var  array
```
```suggestion
	 * @var array
```

eri-trabiccolo commented on includes/abstracts/llms-abstract-query.php:
```diff
@@ -0,0 +1,589 @@
+	/**
+	 * Constructor.
+	 *
+	 * @since 3.8.0
```
```suggestion
	 * @since [version]
```

eri-trabiccolo commented on includes/abstracts/llms-abstract-query.php:
```diff
@@ -0,0 +1,589 @@
+	 *
+	 * @since 3.8.0
+	 *
+	 * @param array $args Optional. Query arguments. Default empty array.
+	 *                    When not provided the default arguments will be used.
```
```suggestion
	 * @param array $args Query arguments. When not provided the default arguments will be used.
```
We decided to drop this optional thing right?

eri-trabiccolo commented on includes/abstracts/llms-abstract-query.php:
```diff
@@ -0,0 +1,589 @@
+	}
+
+	/**
+	 * Get the
```
this suspense is going to kill me

eri-trabiccolo commented on includes/abstracts/llms-abstract-query.php:
```diff
@@ -0,0 +1,589 @@
+		/**
+		 * Filters the allowed sort fields.
+		 *
+		 * The dynamic portion of this hook, `$this->type`, refers to the award type
+		 * for the query, either "achievements" or "certificates".
```
This is not for award types only anymore.
```suggestion
		 * The dynamic part of the filter `$this->id` identifies the extending query.
```

eri-trabiccolo commented on includes/abstracts/llms-abstract-query.php:
```diff
@@ -0,0 +1,589 @@
+
+		/**
+		 * Filters the query default args.
+		 *
```
Misses “The dynamic portion..."
```suggestion
		 * The dynamic part of the filter `$this->id` identifies the extending query.
		 *
```

eri-trabiccolo commented on includes/class-llms-awards-query.php:
```diff
@@ -0,0 +1,340 @@
+	protected $id = 'awards';
+
+	/**
+	 * Specify the post types allowed to be queried by this class
```
```suggestion
	 * Specify the post types allowed to be queried by this class.
```

eri-trabiccolo commented on includes/class-llms-awards-query.php:
```diff
@@ -0,0 +1,340 @@
+	}
+
+	/**
+	 * Retrieve an array of award objects for the given result set returned by the query
```
```suggestion
	 * Retrieve an array of award objects for the given result set returned by the query.
```

eri-trabiccolo commented on includes/class-llms-awards-query.php:
```diff
@@ -0,0 +1,340 @@
+		// Add meta query.
+		$args['meta_query'] = $this->prepare_meta_query();
+
+		foreach ( $args as $arg => $val ) {
+			if ( is_array( $val ) && empty( $val ) ) {
+				unset( $args[ $arg ] );
+			}
+		}
+
+		// Remove empty arrays.
+		return array_filter(
+			$args,
+			function( $val ) {
+				return ! is_array( $val ) || ! empty( $val );
+			}
+		);
```
these two ( the `for` and the `array_filter` ) are equivalent, right?

thomasplevy replied:
yea i forgot to delete the `foreach` after i rewrote it so you'd be impressed with my golfing skills...

thomasplevy committed:
Revert "update default images, per #1081"

This reverts commit a59feb6ad55f85663289314917d229e014dae4a4.

Commit:
Apply suggestions from code review

Co-authored-by: Rocco Aliberti <raliberti84@gmail.com>

thomasplevy committed:
Merge branch 'dev' of github.com:gocodebox/lifterlms into awards-query

Commit:
Apply suggestions from code review

Co-authored-by: Rocco Aliberti <raliberti84@gmail.com>

thomasplevy committed:
add version update

thomasplevy committed:
Remove duplicate code

thomasplevy committed:
Merge branch 'awards-query' of github.com:gocodebox/lifterlms into awards-query

thomasplevy committed:
Merge branch 'engagement-refactoring' of github.com:gocodebox/lifterlms into awards-query

thomasplevy committed:
move award-related methods into a trait

thomasplevy committed:
use counting methods

thomasplevy committed:
don't use deprecated methods

thomasplevy committed:
improve achievement image src getter

eri-trabiccolo commented on includes/admin/reporting/tables/llms.table.students.php:
```diff
@@ -124,7 +126,7 @@ public function get_data( $key, $student ) {
 					)
 				);
-				$value = '<a href="' . esc_url( $url ) . '">' . count( $student->get_achievements() ) . '</a>';
+				$value = '<a href="' . esc_url( $url ) . '">' . $student->get_awards_count( 'achievement' )  . '</a>';
```
```suggestion
				$value = '<a href="' . esc_url( $url ) . '">' . $student->get_awards_count( 'achievement' ) . '</a>';
```

eri-trabiccolo commented on includes/admin/reporting/tables/llms.table.students.php:
```diff
@@ -134,7 +136,7 @@ public function get_data( $key, $student ) {
 					)
 				);
-				$value = '<a href="' . esc_url( $url ) . '">' . count( $student->get_certificates() ) . '</a>';
+				$value = '<a href="' . esc_url( $url ) . '">' . $student->get_awards_count( 'certificate' )  . '</a>';
```
```suggestion
				$value = '<a href="' . esc_url( $url ) . '">' . $student->get_awards_count( 'certificate' ) . '</a>';
```

eri-trabiccolo commented on includes/traits/llms-trait-student-awards.php:
```diff
@@ -0,0 +1,163 @@
+ * Retrieve data related to awards earned by a student
+ *
+ * This trait should only be used by classes that extend from the {@see LLMS_Abstract_User_Data} class.
+ * 
```
```suggestion
 *
```

eri-trabiccolo commented on includes/traits/llms-trait-student-awards.php:
```diff
@@ -0,0 +1,163 @@
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Retrieve data related to awards earned by a student
```
```suggestion
 * Retrieve data related to awards earned by a student.
```

eri-trabiccolo commented on includes/traits/llms-trait-student-awards.php:
```diff
@@ -0,0 +1,163 @@
+<?php
+/**
+ * LifterLMS singleton trait
```
```suggestion
 * LifterLMS singleton trait.
```

eri-trabiccolo commented on includes/traits/llms-trait-student-awards.php:
```diff
@@ -0,0 +1,163 @@
+trait LLMS_Trait_Student_Awards {
+
+	/**
+	 * Retrieve achievements that a user has earned
```
```suggestion
	 * Retrieve achievements that a user has earned.
```

eri-trabiccolo commented on includes/traits/llms-trait-student-awards.php:
```diff
@@ -0,0 +1,163 @@
+	 *
+	 * @since 2.4.0
+	 * @since 3.14.1 Unknown.
+ 	 * @since [version] Moved from `LLMS_Student` class.
```
```suggestion
	 * @since [version] Moved from `LLMS_Student` class.
```

eri-trabiccolo commented on includes/traits/llms-trait-student-awards.php:
```diff
@@ -0,0 +1,163 @@
+
+		$query = $this->get_awards( array(
+			'per_page' => 1,
+			'fields' => 'ids'
```
```suggestion
			'fields' => 'ids',
```

eri-trabiccolo reviewed (changes requested):
minor cs.
Also, are you going to update the `LLMS_Engagement_Handler::create()` in this PR?
I mean b/c now the template id must be "inserted" as the post `parent_id`.

thomasplevy committed:
Use post_parent instead of meta to store template id

thomasplevy committed:
Don't use deprecated methods

thomasplevy committed:
Dont access properties directly

thomasplevy committed:
Switch to parent instead of metadata for storing template

thomasplevy committed:
Use awards query for dupchecking

thomasplevy committed:
Update deprecations

thomasplevy committed:
CS

thomasplevy committed:
Replace preprare with prepare, per #859

Commit:
Apply suggestions from code review

Co-authored-by: Rocco Aliberti <raliberti84@gmail.com>

thomasplevy committed:
D'cs

thomasplevy committed:
Merge branch 'awards-query' of github.com:gocodebox/lifterlms into awards-query
